File: cpp/98-Validate-Binary-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
void check(TreeNode* root,vector<int>& v)
 {
    if(!root) return;
 check(root->left,v);
    v.push_back(root->val);
    check(root->right,v);
}
class Solution {
public:
    bool isValidBST(TreeNode* root) {
     vector<int> v;
    check(root,v);
        for(int i=0;i<v.size()-1;i++)
        {
            if(v[i]>=v[i+1])
                return 0;
        }
        return 1;
    }
};

File: go/BinaryTree/98_Validate_Binary_Search_Tree/98_Validate_Binary_Search_Tree.go
package validatebinarysearchtree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	max := 1 << 32
	left := -1 * max
	right := max
	return isValidBSTUtil(root, left, right)
}

func isValidBSTUtil(root *TreeNode, left int, right int) bool {
	if root == nil {
		return true
	}

	if !(root.Val > left && root.Val < right) {
		return false
	}
	return isValidBSTUtil(root.Left, left, root.Val) && isValidBSTUtil(root.Right, root.Val, right)
}
